### ISSUE TITLE
Continuations and TCO

### ISSUE THREAD

**Issue by jcubic** [jcubic/lips#127]
Implement first class continuations (call/cc) and Tail Call Optimization.

TODO:
- [x] TCO
- [x] call/cc
- [x] lambda continuation
- [x] automatic promise resolving
- [x] Stack trace (continuation based)
- [x] generators
- [x] all let macros
- [x] do and while
- [x] syntax extensions
- [x] load
- [x] quote promise (move macro to js)
- [x] ignore
- [x] bind / unbind
- [x] box / unbox
- [x] syntax-rules
- [x] apply
- [x] higher order functions + escape/restart continuations (move to scheme)
  - [x] map
  - [x] for-each
  - [x] filter
  - [x] reduce
  - [x] fold
  - [x] find
- [x] try..catch
- [x] quasiquote
- [x] macro for ``(let () `(define x ...))``
- [x] syntax extensions
  - [x] `-->`
  - [x] `#`
  - [x] `&`
  - [x] continuation in syntax extension test
- [x] macroexpand
- [x] dynamic scope
- [x] loading bootstrap.scm
- [x] standard library + REPL
- [x] automated tests
  - [x] Core
  - [x] Strings
  - [x] simulate_TCO
  - [x] scope
  - [x] quotation
  - [x] macro
  - [x] syntax-parameters

Documentation
- [x] Update roadmap in README
- [x] delete info that call/cc and TCO are not supported
- [x] add a note about them into lips core
- [x] give credit to [jsScheme](https://bluishcoder.co.nz/jsscheme/)
- [x] blog post

To test the working code (temporary tests):

```bash
git checkout continuations-2
./bin/lips.js --bootstrap none asserts.scm
```

**Comment by jcubic** [comment]
Great post on StackOverflow about TCO and Continuations, it's multiple answers by the same person:

[How to adapt trampolines to Continuation Passing Style?](https://stackoverflow.com/q/57733363/387194)

**Comment by mark-friedman** [comment]
What is the plan with regard to continuations and tail call optimization? In the short term TCO is more important to me, since it is a fairly basic Scheme feature, especially in educational contexts. In the long term it would also be really nice to have call/cc, since there are a bunch of nice Scheme libraries and SRFIs which use it.  It looks like there was some work done on a "continuations" branch back in August of 2021 which was not completed.  Was that because you came across the StackOverflow post that you mention above and decided that that was a better approach?

**Comment by jcubic** [comment]
No, the code in 2021 was based on one implementation, I was not able to make it work. But later I found out that the implementation I based my code on was not actually proper continuations. Even [BiwaScheme don't have 100% working continuation](https://github.com/biwascheme/biwascheme/issues/257).

So now I need to write those from scratch, I started working on it actually in continuations-2 branch. And this time I will base the code on different scheme implementation that do work. The code will be based on [js-scheme](https://bluishcoder.co.nz/jsscheme/) that have most core features of Scheme and the code is small and simple.

I also want to have call/cc and TCO. I planned to do this in a near, unknown future, but now you just give me extra motivation to actually continue working on it. It would be really nice if all examples in [Scheme tutorial](https://lips.js.org/docs/category/introduction-to-scheme) on the website work in LIPS.

**Comment by lassik** [comment]
Worth reading [what IronScheme did](https://github.com/IronScheme/IronScheme/discussions/85) on the .NET framework.

In particular:

> There is a call/cc in IronScheme, but it is not a complete implementation. You can only do escape continuations, basically call/ec from Racket.
> [...]
> The only way to implement call/cc completely in .NET is to use CPS in the compiler. I did such an expirement before, and while it does work, it was horribly slow (10 times or more), and all the .NET interop pretty much goes out of the window.

**Comment by jcubic** [comment]
LIPS already has performance issues. Maybe adding call/cc will not make it slower. Also, in my opinion (because of performance) the best use case for it is in education.

**Comment by mark-friedman** [comment]
> LIPS already has performance issues. Maybe adding call/cc will not make it slower. Also, in my opinion (because of performance) the best use case for it is in education.

Indeed, education is my use case.

**Comment by mark-friedman** [comment]
BTW, I've posted a few messages to your [LIPS gitter](https://gitter.im/jcubic/lips).  Is that the best place to have LIPS related discussions or would you rather do it via some other place, (like, perhaps, GitHub Discussions)?

**Comment by jcubic** [comment]
I'm not at home right now, and I didn't have Gitter tab open.

**Comment by mark-friedman** [comment]
> I also want to have call/cc and TCO. I planned to do this in a near, unknown future, but now you just give me extra motivation to actually continue working on it.

Great!  Thanks.

**Comment by jcubic** [comment]
Whatever works best for you. When I'm at home I will probably reply faster on Gitter and even if I replay late after a few exchanges of messages we will sync and exchange instantly because we will be both live.

**Comment by jcubic** [comment]
Answer to [my question that suggest spaghetti stack](https://stackoverflow.com/a/56482293/387194).

**Comment by mark-friedman** [comment]
Your last comment led me down a bit of a rabbit hole which led me to a few papers that I think might be useful for you.

- [First-Class Continuations on the Java Virtual Machine: An Implementation within the Kawa Scheme Compiler](https://andrebask.github.io/thesis/): It's in the context of Java, rather than JavaScript, but the challenges are very similar.
- [Continuations from Generalized Stack Inspection](https://cs.brown.edu/~sk/Publications/Papers/Published/pcmkf-cont-from-gen-stack-insp/paper.pdf): This paper presents the basis for much of the previous work above.
- [Exceptional Continuations in JavaScript](https://www.schemeworkshop.org/2007/procPaper4.pdf): Some interesting JavaScript-specific work on continuations.
- [unwinder](https://github.com/jlongster/unwinder): Source code for an implementation of the work in "Exceptional Continuations in JavaScript"
- [Implementing a Stepping Debugger in JavaScript](https://archive.jlongster.com/Implementing-Stepping-Debugger-JavaScript): A description of the implementation of unwinder

**Comment by jcubic** [comment]
Thanks.

**Comment by mark-friedman** [comment]
[Compiling for Multi-language Task Migration](https://www-labs.iro.umontreal.ca/~feeley/papers/FeeleyDLS15.pdf): You wouldn't know it by the title, but this is another, potentially useful, paper with info about implementing TCO and continuations for Schemes that interact with other non-Scheme languages.

**Comment by mark-friedman** [comment]
Just curious if this (especially TCO) is something that is being actively worked on (i.e. might be appearing in the relatively near future)?

**Comment by jcubic** [comment]
No, it's not actively worked on right now. I would like to have this implemented, but I don't have much time right now.

**Comment by mark-friedman** [comment]
Thanks for the update, @jcubic. I will plan accordingly.

**Comment by jcubic** [comment]
Started working on them. I'm reading the code of [js-scheme](https://bluishcoder.co.nz/jsscheme/), already found a bug. Let's hope that it will work for my implmentation.

I used simple code from [cse341](https://courses.cs.washington.edu/courses/cse341/04wi/lectures/15-scheme-continuations.html). The the interpreter run into infinite loop when I change the external function into a lambda.

```scheme
(define x ())
(define (put-cont-in-x cont) (set! x cont))
(define print-line (lambda (x) (display x) (newline)))

(print-line (call/cc put-cont-in-x))
;; ==> #<void>
(x "hi!")
;; ==> "hi"
```

Above code works fine, but replacing `put-cont-in-x` with lambda crash the browser tab.

**Comment by mark-friedman** [comment]
> Above code works fine, but replacing `put-cont-in-x` with lambda crash the browser tab.

Hm, it worked just fine for me when I used the `lambda` version in[ js-scheme](https://bluishcoder.co.nz/jsscheme/).  Here's the log:
```scheme
(define x ())
=> ()
(define (put-cont-in-x cont) (set! x cont))
=> #lambda
(define print-line (lambda (x) (display x) (newline)))
=> #lambda
(print-line (call/cc (lambda (cont) (set! x cont))))
#lambda
=> #null
(x "hi!")
hi!
=> #null
```
I'm using Chrome version 132.0.6834.83 on a Mac.

**Comment by jcubic** [comment]
Yeah, it also work for me. Maybe I did something wrong with the code.

**Comment by jcubic** [comment]
The base code of TCO and continuation is working based on [js-scheme](https://bluishcoder.co.nz/jsscheme/) by Alex Yakovlev. I've created a function `tco_eval` that I test inside JavaScript. Now I need to refactor the rest of the code and add missing features from old evaluate. The hard part is done.

**Comment by mark-friedman** [comment]
> The base code of TCO and continuation is working ... The hard part is done.

That's great to hear!

>Now I need to refactor the rest of the code and add missing features from old evaluate.

Good luck!

**Comment by jcubic** [comment]
It was too soon, I spend the whole day to fix simple loop with continuations, but it's working now:

```scheme
(assert (let ((i 10) (result ()))
          ((lambda (loop)
            (if (eq? i 0)
                result
                (begin
                  (set! i (- i 1))
                  (set! result (cons i result))
                  (loop loop))))
           (call/cc (lambda (cc) cc))))
   (list 0 1 2 3 4 5 6 7 8 9))
```

Will keep adding more assertions until I will be able to run normal tests. Assert is a macro:

```scheme
(define-macro (assert a b)
   (let ((a_expr (gensym "a"))
         (b_expr (gensym "b")))
     `(let ((,a_expr (repr ,a))
            (,b_expr (repr ,b)))
          (if (not (eq? (--> ,a_expr (cmp ,b_expr)) 0))
              (print (concat "FAIL: "
                             ,a_expr
                             " != "
                             ,b_expr))))))
```

**Comment by mark-friedman** [comment]
> fix for Petrofsky catastrophe

Wow! This led me down a major USENET rabbit hole!  [This](https://groups.google.com/g/comp.lang.scheme/c/J2ut9mbyQrU/m/UbZCSoA4XhQJ) was the result.

**Comment by jcubic** [comment]
Wow, you actually found it. The link on scheme.org was dead, and it was the only search result.

**Comment by jcubic** [comment]
You should create a PR and update:

https://docs.scheme.org/surveys/petrofsky-catastrophe/

The link to github is on the page.

**Comment by mark-friedman** [comment]
> You should create a PR and update:
> 
> https://docs.scheme.org/surveys/petrofsky-catastrophe/
> 
> The link to github is on the page.

https://github.com/schemedoc/surveys/pull/69

**Comment by mark-friedman** [comment]
> all lambda functions are now async

Are there any performance implications of this?

Also, does this mean that we'll have to explicitly `await` lambda functionsin our JavaScript code or is that done implicitly via the Scheme<->JavaScript interop?

**Comment by jcubic** [comment]
In JavaScript, you will need to explicitly use await. It may be problematic when using callback based code. But lambda could be async anyway when you use async macro or function written in JavaScript.

Now when I think about this, I should add warning about functions like Array::filter that will not work, right now it can break randomly, when lambda become async.

I didn't test the performance, but I think that the most things that slow down LIPS are not related to interpretation, only interop with JavaScript. The lot of stuff is happening with boxing/unboxing LIPS values and binding/unbinding functions (LIPS is what I call soft bind store the original function so you can unbind it and get original unbound function, I'm talking bout fn.bind in JavaScript that change the context of `this`).

**Comment by mark-friedman** [comment]
> In JavaScript, you will need to explicitly use await. It may be problematic when using callback based code. But lambda could be async anyway when you use async macro or function written in JavaScript.

Callbacks and collection type code (like Array::filter) are fairly important on the JavaScript side. How about doing something similar to what you do on the Scheme side for automagic async/await and promise quotation?

**Comment by jcubic** [comment]
I was just about posting that's not possible, but it get me thinking. It will require additional work because you will need to wrap the native function with another function, that will sit in between Scheme and that function. It may even work. I need to think about this.

**Comment by jcubic** [comment]
I don't think that this is possible. You will need to execute the function twice (so lambda can't have side effects), or create a new implementation for every function that is async.

I did something like this for [Gaiman](https://github.com/jcubic/gaiman), The functions are also async there.

I created a new class that extends Array and rewrite every method:

https://github.com/jcubic/gaiman/blob/6b9656515d9f045359b3eec72ea9787d625afc56/src/prefix.js#L384-L397

**Comment by mark-friedman** [comment]
It would be very unfortunate if LIPS lost the basic ability to pass ordinary Scheme procedures (that don't explicitly return promises) to JavaScript functions and have them be treated as if they return promises.

**Comment by jcubic** [comment]
The new interpreter to handle promises needs to be async because it's a while loop, and since lambda call interpreter (when called from JavaScript) it has to return a promise.

To keep the old behavior I would need to keep the lambda use old interpreter and be very limited. No Continuations and TCO.

Lambdas right now are handled in two ways, when called from LIPS code it use the core of the interpreter, but when you use lambda outside of LIPS it's a function that call an interpreter inside. I yet need to play with lambda called from JavaScript to see if there are limitations to continuations.

Now I'm thinking, I don't like this idea very much, but If I have two interpreters one async and one sync, I would be able to use sync interpreter in lambda, that code is not used when you lambda as value in JavaScript.

Also, another note is that async..await can be abstracted by generators and yield. This is what babel was doing when transpiling async await for older JavaScript. I can try to rewrite the code to use generators instead of async await, and see it will return a Promise or a value. This is another thing that I can create a PoC for.

**Comment by jcubic** [comment]
Wow, this actually works:

```scheme
(print (let ((l (list->array (list 1 2 3 4))))
         (l.filter (lambda (x) (>= x 3)))))
;; ==> [3,4]
```

Got this result, but my previous code that I was testing stopped working.

**Comment by jcubic** [comment]
All my temporary tests are passing. It looks like none of the core of LIPS will be async.

**Comment by mark-friedman** [comment]
Is the continuations-2 branch at the point yet where I can build it and run a standalone REPL (as well as the various JS libraries)?  If so, how would I do that?

**Comment by jcubic** [comment]
There is still a lot of work. Probably the biggest thing is quasiquote that need to use new interpreter, REPL still use the old code. And I still use temporary tests that I run to see if I don't break anything. It may take a while before it's finished.

**Comment by mark-friedman** [comment]
> There is still a lot of work. Probably the biggest thing is quasiquote that need to use new interpreter, REPL still use the old code. And I still use temporary tests that I run to see if I don't break anything. It may take a while before it's finished.

I see. Well, I am looking forward to trying it out (even if it has some limitations) once it is at some kind of semi-usable state, so let me know.

Great work!  Keep it up.

**Comment by mark-friedman** [comment]
BTW, if you need some more tests you can try the ones [here](https://gitlab.com/kashell/Kawa/-/blob/callcc/testsuite/cont-test.scm), from a [branch of Kawa that added full continuations](https://gitlab.com/kashell/Kawa/-/blob/callcc).  That Kawa test file also references [these](https://web.archive.org/web/20230320180805/http://rigaux.org/language-study/various/callcc/scheme.listing) continuation tests.

**Comment by jcubic** [comment]
I can add my playground code to the repository in the form of my asserts. If you want to play. But note that you can't use Promise quotations, and lisp macros that are in a form ``(let () `(define xxx something))``. the old quasiquote get let as scope for the code and `xxx` is undefined when evaluating the macros.

Also I didn't tested syntax-rules that also need to be refactored. But that one should be simple since syntax-rules is only symbols swapping that get evaluated inside JavaScript macro at the end. But you can't use `define-syntax` because it's a macro that use ``(let ((... (gensym))) `(define ,name code)`` + some type checking.

I will also add my todo list of the issue, so it's easier to see how much work need to be done.

**Comment by jcubic** [comment]
There is a problem with code that replaced async/await with generators. I was using bun to test the code the whole time, and it has a bigger stack size.

When running my testing code on Node.js got this error:

```
RangeError: Maximum call stack size exceeded at line 33
    at tco_generator (file:///home/kuba/projects/jcubic/scheme/lips/continuations/src/lips.js:11417:31)
    at tco_generator.next (<anonymous>)
    at next (file:///home/kuba/projects/jcubic/scheme/lips/continuations/src/lips.js:3916:33)
    at unpromise (file:///home/kuba/projects/jcubic/scheme/lips/continuations/src/lips.js:1948:12)
    at next (file:///home/kuba/projects/jcubic/scheme/lips/continuations/src/lips.js:3920:16)
    at unpromise (file:///home/kuba/projects/jcubic/scheme/lips/continuations/src/lips.js:1948:12)
    at next (file:///home/kuba/projects/jcubic/scheme/lips/continuations/src/lips.js:3920:16)
    at unpromise (file:///home/kuba/projects/jcubic/scheme/lips/continuations/src/lips.js:1948:12)
    at next (file:///home/kuba/projects/jcubic/scheme/lips/continuations/src/lips.js:3920:16)
    at unpromise (file:///home/kuba/projects/jcubic/scheme/lips/continuations/src/lips.js:1948:12)
```

Also, the REPL is working, but you need to execute with:

```
./bin/lips.js --bootstrap none -q
```

**Comment by jcubic** [comment]
It should work now. You can use REPL without standard library and TCO and continuations works.

**Comment by jcubic** [comment]
I think that I've finally found the last bug in continuations, the reason why generator with `for-each` was not working.

The problem was `apply` that use JavaScript based code for the lambda.

This will be a limitation of continuations, when you pass a function to JavaScript as callback you can't use continuations that reach outside of the function, since you leave the Scheme realm, and lambda become function that evaluate the code with top level continuation.

`apply` was working like this, it needs to be part of the interpreter core.

**Comment by jcubic** [comment]
`ignore` will also work in similar way, it needs to be called with top level continuation, so also ignore continuations that are outside.

In both cases, using continuations on boundary of JavaScript and `ignore` will be unspecified (similarly as how R<sup>n</sup>RS specify some behavior).

**Comment by jcubic** [comment]
I found that js-scheme don't actually have TCO, the recursive functions never overflow.

Both functions work exactly the same:

```scheme
(define (sum n)
  (if (<= n 0)
      0
      (+ n (sum (- n 1)))))

(define (sum n)
  (let loop ((n n) (result 0))
    (if (<= n 0)
        result
        (loop (- n 1) (+ n result)))))
```

**Comment by mark-friedman** [comment]
> I found that js-scheme don't actually have TCO, the recursive functions never overflow.

I just profiled jsScheme on my browser and saw that the non-tail-call version of `sum` did use up increasing amounts of **heap** memory, while the tail-call version of `sum`did not.  I believe that that still counts as TCO (which does not refer only to stack overflow, but to memory use in general).

**Comment by jcubic** [comment]
Yeah, non TCO will create very long Continuation chain, but it don't overflow. Right now, computers have so much memory, so it will not run out of memory with simple test cases.

**Comment by mark-friedman** [comment]
Maybe you could set the V8 heap size in Node to a low level during testing.  See https://nodejs.org/docs/latest/api/cli.html#--max-old-space-sizesize-in-mib

**Comment by mark-friedman** [comment]
I know that you're still working on this code, so maybe this is premature and maybe you already know about this.  I am seeing what looks like non-linear time behavior and linear growing space behavior on what should be TCO loops.   The LIPS code that I'm using to test this is:
```
lips> (define (myloop x) (if (<= x 0) 'done (myloop (- x 1))))
lips> (define (timed-loop x) (begin (globalThis.console.time "foo") (myloop x)  (globalThis.console.timeEnd "foo")))
lips> (timed-loop 200000)      ; heap:764MB
foo: 3:10.458 (m:ss.mmm)
lips> (timed-loop 100000)       ; heap:411MB
foo: 49.613s
lips> (timed-loop 50000)         ; heap:253MB
foo: 13.817s
lips> (timed-loop 25000)         ; heap:170MB
foo: 3.875s
```
I've appended each loop eval with the amount of memory it used at its maximum.  The baseline memory use (i.e. when the loop wasn't running) was about 55MB.  You can see roughly quadratic growth in time and linear growth in memory.  The memory usage is based on observing my Mac's Activity Monitor. 

I've run the test a number of times and in different orders and see pretty much the same thing.

**Comment by mark-friedman** [comment]
> Each loop will create a new continuation, so the memory should grow, but it should be garbage collected. 

But it doesn't seem to get GC-ed during the execution of the loop, though, so the effective behavior is that the TCO loop ends up using up memory as if it wasn't TCO.  And there is still the issue of the non-linear time increases, which is somewhat strange, though perhaps it's related to the memory issues in some way.

> There are already [issues related to performance](https://github.com/jcubic/lips/issues?q=is%3Aissue%20state%3Aopen%20speed).

Understood.  Though the issues I'm seeing might be new issues.

> I will try to use some profiler, maybe will find what is causing performance and memory for such simple code.

Sounds good!  Thanks!

**Comment by jcubic** [comment]
I've tested this code in the browser:

```html
<script type="module" src="./dist/lips.js"></script>
<script type="text/x-scheme">
(define (sum n acc)
  (if (<= n 0)
      acc
      (sum (- n 1) (+ acc n))))

(set! window.onload
      (lambda ()
         (let ((node (document.querySelector "button")))
           (node.addEventListener "click"
                                  (lambda ()
                                    (print (sum 20000 0)))))))
</script>
<button>run</button>
```

And the initial memory 10MB and during the execution of the function the size increased to 11MB.
I've also run profiler and can see Minor GC run during the execution of the code.

Maybe will try to hook up Node process to browser dev tools. Never done this before.

**Comment by jcubic** [comment]
I was wrong, the size increased from 10MB to 49MB and then dropped after the code finish.

**Comment by mark-friedman** [comment]
What are you seeing in terms of time and max memory as you change the number of iterations of the loop?  Are you seeing the same kind of (almost) quadratic time increases and linear memory increases that I saw on Node?

**Comment by jcubic** [comment]
Memory increase is linear, not sure about the time, but it's probably the same. Both Brave browser and Node.js use V8 engine. Only the memory usage is smaller.

I'm not sure how to optimize the numbers to see if there are any difference.

I will check js-scheme how it performs, but the code is way faster, so it may be harder to measure.

**Comment by jcubic** [comment]
I created a quick test:

```scheme
(define one 1)
(define zero 0)

(define (myloop x)
  (if (<= x zero)
      'done
      (begin
        (x.dec one)
        (myloop x))))
```

I've added mutation method on LNumber:

```javascript
LNumber.prototype.dec = function(n) {
    this.constant(this.__value__ - n.__value__, this.__type__);
};
```

And there are no memory increasing. So for every recursive call, 3 instance of numbers were created. `0` `1` and `x--`.

I can make the Number what I did with symbols, they keep a cache of previous symbols created. So they are so called interned symbols. This will solve the issue with `0` and `1`, but still there is a problem of create new number for the x. Caching the value will not help since every number is different.

I was thinking of something like reusing of symbols that are not used anymore, but I have no idea how to pull this off.

I will try to ask on Reddit.

**Comment by mark-friedman** [comment]
I added your mutation and ran your loop with various values of x and found the same sort of quadratic time behaviors as before (i.e. doubling x will cause a ~4X increase in the time it takes for the loop).  So it looks like even if the memory issue is dealt with, the time issue will remain.

Also note that in addition to adding the mutation method, I had to change the definition of the `enumerable` function to add `writable: true`.  Otherwise I received an error:
```
TypeError: Cannot redefine property: __value__ at line 1
    at Function.defineProperty (<anonymous>)
    at enumerable (file:///Users/mark/code/jcubic/lips/src/lips.js:5472:12)
    at LNumber.constant (file:///Users/mark/code/jcubic/lips/src/lips.js:5967:5)
    at LNumber.dec (file:///Users/mark/code/jcubic/lips/src/lips.js:5972:10)
    at call_function (file:///Users/mark/code/jcubic/lips/src/lips.js:11224:32)
    ...
```
Perhaps that's just a Node problem, though.

**Comment by jcubic** [comment]
Yes, I forget that I needed to modify `enumerable`.

There is also an issue of environment allocation, that someone pointed on Reddit, where I asked about this. I need to somehow figure out how to reuse the environment on recursive call.

**Comment by mark-friedman** [comment]
> There is also an issue of environment allocation, that someone pointed on Reddit, where I asked about this. I need to somehow figure out how to reuse the environment on recursive call.

Makes sense. It's not clear that that will fix the quadratic time behavior, though, but perhaps in figuring out how to reuse the environment you will also realize what is causing the quadratic time behavior.  I'm also starting to take a closer look at the code, so maybe I'll be able to help.

**Comment by jcubic** [comment]
I've cached the environment created by the call_function.

So this code:

```
(define (myLoop x)
  (let ((x x))
    (letrec ((loop (lambda ()
                    (if (> x zero)
                        (begin
                          (set! x (- x one))
                          (loop))))))
      (loop))))
```

when runs:
* 100000 - 14.81s
* 200000 - 24.87s
* 300000 - 45.27s
* 400000 - 54.21s


```mermaid
xychart-beta
    x-axis "loops" [10000, 50000, 100000, 200000, 300000, 400000]
    y-axis "time (s)" 0 --> 60
    line [1.29528, 6.67, 14.81, 24.87, 45.27, 54.21]
```

The memory is still increasing

**Comment by mark-friedman** [comment]
I've had a little bit of time to profile the memory use of the continuations-2 branch, using Chrome's dev tools.  By taking some heap snapshots you can see the increase in memory as tail call loops run.  In particular, I notice that the memory is largely created in the following LIPS functions/classes:

- Environment
- Continuation 
- State

My guess is that the interpreter is not reusing those objects when performing tail calls, though I don't have enough understanding of the LIPS code to verify that guess.

**Comment by jcubic** [comment]
I'm not sure if you can eliminate Continuation class that is created for every pair in the code.
State should not be created because there is only one instance, unless you execute function that use `tco_eval`, they should be eliminated when the feature is done.

Environment in lambda is eliminated, but you can't eliminate the lambda in let that is inside lambda, at least I don't know how to do this.

You basically would need to have what JIT compilation have in order to optimize it. I have no idea how to implement such a thing.

**Comment by mark-friedman** [comment]
> I'm not sure if you can eliminate Continuation class that is created for every pair in the code.

I'm not entirely sure what that means.  What I am saying is that during evaluation tail calls should be able to re-use the dynamic continuation and, I think, the environment, at that point rather than creating a new one.  I believe that that's the essential nature of tail call optimization.

**Comment by mark-friedman** [comment]
@jcubic Have you looked back at the Stack Overflow articles about trampolines that you posted a couple of years ago? Maybe that could help you.
[How to adapt trampolines to Continuation Passing Style?](https://stackoverflow.com/q/57733363/387194)

**Comment by jcubic** [comment]
They will be no helpful, Trampoline is to implement stack less recursion (TCO), it will not help with memory and time usage.

I didn't work on this for a while, but I think that if I want to have continuations and TCO, I need to accept the limitations and memory usage. I have no idea how to optimize the code to not consume the resources, maybe I will need to have something like JIT that compile the code into JavaScript. This is what jsScheme was doing, I'm not understood how that part of the code is working, but it would be not possible to compile then code like this. LIPS have too many runtime manipulations to support interop with JavaScript.

**Comment by mark-friedman** [comment]
Good point about the trampolines.  I guess they won't help in this situation.  

>I think that if I want to have continuations and TCO, I need to accept the limitations and memory usage.

I hate to say it, but the memory usage might make the solution not really satisfy the conditions of TCO (which are not just about stack usage).  That said, I guess it's better than blowing the JavaScript stack.

I wish I understood your code well enough to offer some help.

**Comment by jcubic** [comment]
Started working on this again the other way with help from AI:


```mermaid
xychart-beta
    title "Duration in seconds for a given loop"
    x-axis "loops" [10000, 20000, 50000, 100000, 200000, 300000, 400000]
    y-axis "time (s)" 0 --> 20
    line [0.13599, 0.263931, 0.629128, 1.269, 2.549, 3.913, 4.973]
```


```mermaid
xychart-beta
    title "LIPS heap usage during a 200k tail-call loop (heapUsed, MB)"
    x-axis "eval steps (thousands)" 0 --> 270
    y-axis "heap used (MB)" 0 --> 45
    line [11.4, 16.8, 22.2, 27.5, 32.9, 38.2, 11.5, 16.7, 21.9, 27, 32.2, 37.4, 42.6, 15.8, 20.9, 26, 31.1, 36.1, 41.2, 14.3, 19.3, 24.4, 29.5, 34.5, 39.6, 12.7, 17.7, 22.8, 27.9, 33, 38, 43.1, 16.2, 21.2, 26.3, 31.4, 36.4, 41.5, 14.6, 19.6, 24.7, 29.7, 34.7, 39.7, 12.7, 17.7, 22.7, 27.6, 32.6, 37.6, 42.6, 15.5, 20.5, 25.4, 30.3, 35.3, 40.2, 13.1, 18, 22.9, 27.9, 32.8, 37.7, 42.6, 15.5, 20.4, 25.3, 30.2, 35.1, 39.9, 12.8, 17.7, 22.6, 27.5, 32.4, 37.2, 42.1, 15, 19.9, 24.8, 29.7, 34.5, 39.4, 12.3, 17.2, 22.1, 27, 31.9, 36.7, 41.6]
```
